README.md

# janeaustenr <img src="tools/logo.png" align="right"/>

<!-- badges: start -->
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/janeaustenr)](https://cran.r-project.org/package=janeaustenr)
[![R-CMD-check](https://github.com/juliasilge/janeaustenr/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/juliasilge/janeaustenr/actions/workflows/R-CMD-check.yaml)
[![DOI](https://zenodo.org/badge/22224/juliasilge/janeaustenr.svg)](https://zenodo.org/badge/latestdoi/22224/juliasilge/janeaustenr)
<!-- badges: end -->

## An R Package for Jane Austen's Complete Novels

> “The person, be it gentleman or lady, who has not pleasure in a good novel, 
> must be intolerably stupid.”

(from Mr. Tilney in *Northanger Abbey*)

This package provides access to the full texts of Jane Austen's 6 completed, published novels. The UTF-8 plain text for each novel was sourced from [Project Gutenberg](https://www.gutenberg.org/), processed a bit, and is ready for text analysis. Each text is in a character vector with elements of about 70 characters. The package contains:

* `sensesensibility`:  *Sense and Sensibility*, published in 1811
* `prideprejudice`:  *Pride and Prejudice*, published in 1813
* `mansfieldpark`:  *Mansfield Park*, published in 1814
* `emma`:  *Emma*, published in 1815
* `northangerabbey`:  *Northanger Abbey*, published posthumously in 1818
* `persuasion`:  *Persuasion*, also published posthumously in 1818

There is also a function `austen_books()` that returns a tidy data frame of all 6 novels. 

Users should be aware that there are some differences in usage between the novels as made available by Project Gutenberg. For example, "anything" vs. "any thing", "Mr" vs. "Mr.", and using underscores vs. all caps to indicate italics/emphasis.

## Installation

You can install the released version of janeaustenr from
[CRAN](https://CRAN.R-project.org) with:

```r
install.packages("janeaustenr")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("juliasilge/janeaustenr")
```

## How to Use This Package

For some ideas on getting started with analyzing these texts, see my [blog post on sentiment analysis of Austen's novels](https://juliasilge.com/blog/if-i-loved-nlp-less/). For help within R, try `?persuasion` or similar for getting started with the data sets.

This project is released with a [Contributor Code of
Conduct](https://www.contributor-covenant.org/version/2/1/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# janeaustenr (development version)

# janeaustenr 1.0.0

* Use suggested packages more strictly

# janeaustenr 0.1.5

* Fixed encoding for *Mansfield Park*
* Added package to calls to data objects, since they are lazy-loaded and not in the namespace

# janeaustenr 0.1.4

* Actually fixed factor order in `austen_books` function to align with publication order

# janeaustenr 0.1.3

* Attempted to fix factor order in `austen_books` function to align with publication order (made an error in this)
* Added unit test to check output of `austen_books`

# janeaustenr 0.1.2

* Moved `dplyr` to Suggests; change implementation of `austen_books` to use base functions thanks to Jeroen Ooms

# janeaustenr 0.1.1

* Added a `NEWS.md` file to track changes to the package.
* Added some details on usage differences between novels to README
* Replaced all data files with new versions to solve problem of formatting change at 10000 lines

# janeaustenr 0.1.0

* Initial release of full texts of Jane Austen's 6 completed, published novels


tests/testthat/test_austen_books.R
# tests for austen_books function

all_books_in_order <- c(
  "Sense & Sensibility",
  "Pride & Prejudice",
  "Mansfield Park",
  "Emma",
  "Northanger Abbey",
  "Persuasion"
)

test_that("factor order is correct", {
  book_lines <- austen_books()
  expect_identical(levels(book_lines$book), all_books_in_order)
})

test_that("tidy frame for Austen books is right", {
  skip_if_not_installed("dplyr")
  lines_by_book <- austen_books() %>% 
    dplyr::group_by(book) %>%
    dplyr::summarise(total_lines = dplyr::n())
  expect_identical(nrow(lines_by_book), 6L)
  expect_identical(ncol(lines_by_book), 2L)
  # the factor levels still in the right order?
  expect_identical(
    as.character(lines_by_book$book),
    all_books_in_order
  )
  # Persuasion has fewer lines than Emma?
  expect_lt(
    lines_by_book$total_lines[[6]],
    lines_by_book$total_lines[[4]]
  )
})


R/austen_books.R
#' Tidy data frame of Jane Austen's 6 completed, published novels
#' 
#' Returns a tidy data frame of Jane Austen's 6 completed, published novels with 
#' two columns: `text`, which contains the text of the novels divided into 
#' elements of up to about 70 characters each, and `book`, which contains the titles of
#' the novels as a factor in order of publication.
#' 
#' @details Users should be aware that there are some differences in usage 
#' between the novels as made available by Project Gutenberg. For example, 
#' "anything" vs. "any thing", "Mr" vs. "Mr.", and using underscores vs. all 
#' caps to indicate italics/emphasis.
#' 
#' @return A data frame with two columns: `text` and `book`
#' 
#' @name austen_books
#' 
#' @examplesIf requireNamespace("dplyr", quietly = TRUE)
#' 
#' library(dplyr)
#'
#' austen_books() %>% 
#'     group_by(book) %>%
#'     summarise(total_lines = n())
#'
#' @export
austen_books <- function(){
        books <- list(
                "Sense & Sensibility" = janeaustenr::sensesensibility,
                "Pride & Prejudice" = janeaustenr::prideprejudice,
                "Mansfield Park" = janeaustenr::mansfieldpark,
                "Emma" = janeaustenr::emma,
                "Northanger Abbey" = janeaustenr::northangerabbey,
                "Persuasion" = janeaustenr::persuasion
        )
       ret <- data.frame(text = unlist(books, use.names = FALSE), 
                         stringsAsFactors = FALSE)
       ret$book <- factor(rep(names(books), sapply(books, length)))
       ret$book <- factor(ret$book, levels = unique(ret$book))
       structure(ret, class = c("tbl_df", "tbl", "data.frame"))
}


R/novels.R
#' The text of Jane Austen's novel "Sense and Sensibility"
#'
#' A dataset containing the text of Jane Austen's 1811 novel "Sense and 
#' Sensibility". The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source <http://www.gutenberg.org/ebooks/161>
#' @format A character vector with 12262 elements
"sensesensibility"

#' The text of Jane Austen's novel "Pride and Prejudice"
#'
#' A dataset containing the text of Jane Austen's 1813 novel "Pride and 
#' Prejudice". The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source <http://www.gutenberg.org/ebooks/1342>
#' @format A character vector with 12447 elements
"prideprejudice"

#' The text of Jane Austen's novel "Mansfield Park"
#'
#' A dataset containing the text of Jane Austen's 1814 novel "Mansfield
#' Park". The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source <http://www.gutenberg.org/ebooks/141>
#' @format A character vector with 14768 elements
"mansfieldpark"

#' The text of Jane Austen's novel "Emma"
#'
#' A dataset containing the text of Jane Austen's 1815 novel "Emma".
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each. 
#' (Some elements are blank.)
#'
#' @source <http://www.gutenberg.org/ebooks/158>
#' @format A character vector with 15297 elements
"emma"

#' The text of Jane Austen's novel "Northanger Abbey"
#'
#' A dataset containing the text of Jane Austen's novel "Northanger Abbey", 
#' published posthumously in 1818.
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source <http://www.gutenberg.org/ebooks/121>
#' @format A character vector with 7840 elements
"northangerabbey"

#' The text of Jane Austen's novel "Persuasion"
#'
#' A dataset containing the text of Jane Austen's novel "Persuasion", 
#' published posthumously in 1818.
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source <http://www.gutenberg.org/ebooks/105>
#' @format A character vector with 8328 elements
"persuasion"
